2.0.0 rewrite using Custom Resources

gsaslis opened the pull request:
Allows more than one ghost blogs to be deployed on the same server, through the rewrite of the v1.x recipes as (Chef 12.5) Custom Resources.  

Sample usage can be found in the test cookbook that is used for Integration Tests (using Test Kitchen and Inspec verifier). 

The basic idea is that each ghost blog installation is treated as a separate system service and is independent of the nginx reverse proxy (for which a second custom resource is provided). 

Some other cleanup as well.

Commit:
Indentation fix only

Commit:
Removed Vagrantfile, in favour of simply using Test Kitchen

Further reading: https://sethvargo.com/the-future-of-vagrant-berkshelf/

Commit:
Added new Custom Resource that handles ghost blog installation

This basically does the same as the old _ghost recipe, only it has been made reusable, so it can be used many times on the same server.
Please note:
* Each ghost blog will go in a separate path.
* You will need to define a different port for each ghost blog.
* Each ghost blog has their own separate system service now for managing it.

Commit:
Renamed and cleaned up custom resource

Commit:
Replaced _ghost recipe usage in default recipe with example for using the Custom Resources

Commit:
Converted nginx recipe into custom resource

Commit:
Removed _nginx recipe, as it's been replaced with the custom resource

Commit:
Added example for `ghost_nginx` custom resource

Commit:
Added test cookbook that tests usage of custom resources

Commit:
Bugfix: added cookbook definition to clarify where template files are loaded from

Commit:
Removed ghost service from services recipe, as there is now a separate service per ghost blog

Commit:
Cleanup in nginx custom resource

Commit:
Enable the system services by default, but remove their installation by default.

The default recipe just installs some dependencies and it is up to the user of the cookbook to include the _services recipe if it needs the nginx service explicitily enabled. The _services recipe should be included in the run_list whenever the ghost_ngix resource is used.

Commit:
Fixed incorrect passing of variables to template

Commit:
Fixed test recipe with correct values (matching the blog_example recipe)

Commit:
Added inspec verifier Integration Tests

Commit:
Updated .kitchen.yml to
* use new test cookbook,
* use chef_zero instead of chef_solo,
* removed ubuntu 12.04 support (some weird issue with connecting to the instance)
* showcasing installation of node 6.9.4

Commit:
Added my name to the maintainers

Commit:
Major version bump to indicate significant changes / rewrite / breaking backwards compatibility

Commit:
Added sample inspec attributes file

arukaen commented:
``` sh
Profile: IT tests for Ghost. (ghost-inspec-profile)
Version: 1.0.0
Target:  ssh://vagrant@127.0.0.1:2222

  ✔  Ghost Blog: Ghost Blog Tests
     ✔  File /var/www/html/ghost_1 should be directory
     ✔  File /var/www/html/ghost_1 type should eq :directory
     ✔  File /var/www/html/ghost_1 owner should eq "root"
     ✔  File /var/www/html/ghost_1 group should eq "root"
     ✔  File /etc/init.d/ghost_test content should match "/var/www/html/ghost_1"
     ✔  File /etc/init.d/ghost_test content should match "ghost_test"
     ✔  File /var/www/html/ghost_1/config.js content should match "http://myblog.com"
     ✔  File /var/www/html/ghost_1/config.js content should match "2378"
     ✔  Service ghost_test should be installed
     ✔  Service ghost_test should be enabled
     ✔  Service ghost_test should be running
     ✔  Port 2378 should be listening
  ✔  Ghost Blog Nginx Config: Ghost Nginx Config
     ✔  System Package nginx should be installed
     ✔  Service nginx should be enabled
     ✔  Service nginx should be installed
     ✔  Service nginx should be running
     ✔  File /etc/nginx/sites-available/test.conf content should match "myblog.com"
     ✔  File /etc/nginx/sites-available/test.conf content should match "test"
     ✔  File /etc/nginx/sites-available/test.conf content should match "2378"
     ✔  Port 80 should be listening

Profile Summary: 2 successful, 0 failures, 0 skipped
Test Summary: 20 successful, 0 failures, 0 skipped
```

@gsaslis  tests pass but there are these foodcritic issues:

```
FC038: Invalid resource action: ./libraries/resource_blog.rb:104
FC064: Ensure issues_url is set in metadata: ./metadata.rb:1
FC065: Ensure source_url is set in metadata: ./metadata.rb:1
```

Commit:
Fixed issue with multiple system services shutting each other down.

This was because the `start-stop-daemon`, when using the `--exec` param looks for all processes that are instances of the (in our case) `/usr/local/bin/node` executable, meaning it would essentially kill not just the current blog running, but any process running under node.js.

 This caveat is well-documented in the `start-stop-daemon` man page, where it is stated that it is not a suitable approach for interpreted scripts. This leaves us with the only option to use the `--pidfile` option, so we can be sure we're always just killing what we should be.

Here's the relevant info from the [man page](http://man7.org/linux/man-pages/man8/start-stop-daemon.8.html):
```
-x, --exec executable
              Check for processes that are instances of this executable. The
              executable argument should be an absolute pathname. Note: this
              might not work as intended with interpreted scripts, as the
              executable will point to the interpreter. Take into account
              processes running from inside a chroot will also be matched,
              so other match restrictions might be needed.
 ```

Commit:
Fixed foodcritic warnings

1. Was using invalid action on `service` resource,
2. some metadata info missing

Commit:
Added rakefile to simplify the structured running of all tests.

Commit:
Automated style fixes from running `rubocop -a`

Commit:
Fixed remaining rubocop issues

gsaslis commented:
@arukaen nice catch! 

fixed those and also added rubocop tests and a Rakefile to facilitate running all tests
